File: Codes in Python3/Dynamic Programming/Matrix Chain Multiplication.py
from math import *

class Solution:
    def matrixMultiplication(self, N, arr):
        dp=[[0]*N for _ in range(N)]
        for i in range(N-1,0,-1):
            for j in range(i+1,N):
                dp[i][j]=inf
                for k in range(i,j):
                    dp[i][j]=min(dp[i][j],dp[i][k]+dp[k+1][j]+arr[i-1]*arr[k]*arr[j])
        
        return dp[1][N-1]

File: Codes in C++/Dynamic Programming/Matrix Chain Multiplication.cpp
class Solution
{
public:
    int matrixMultiplication(int N, int arr[])
    {
        vector<vector<int>> mcm(N, vector<int>(N, 0));
        for (int i = N - 1; i >= 1; i--)
        {
            for (int j = i + 1; j < N; j++)
            {

                mcm[i][j] = INT_MAX;
                for (int k = i; k < j; k++)
                {
                    mcm[i][j] = min(mcm[i][j], mcm[i][k] 
                                    + mcm[k + 1][j] 
                                    + arr[i - 1] * arr[k] * arr[j]);
                }
            }
        }
        return mcm[1][N - 1];
    }
};